src/stdlib/Conversions.scillib

```diff
@@ -0,0 +1,221 @@
+(* ******************************************************************************** *)
+(*   This file is part of scilla.                                                   *)
+(*                                                                                  *)
+(*   Copyright (c) 2018 - present Zilliqa Research Pvt. Ltd.                        *)
+(*                                                                                  *)
+(*   scilla is free software: you can redistribute it and/or modify it under the    *)
+(*   terms of the GNU General Public License as published by the Free Software      *)
+(*   Foundation, either version 3 of the License, or (at your option) any later     *)
+(*   version.                                                                       *)
+(*                                                                                  *)
+(*   scilla is distributed in the hope that it will be useful, but WITHOUT ANY      *)
+(*   WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR  *)
+(*   A PARTICULAR PURPOSE.  See the GNU General Public License for more details.    *)
+(*                                                                                  *)
+(*   You should have received a copy of the GNU General Public License along with   *)
+(*   scilla.  If not, see <http://www.gnu.org/licenses/>.                           *)
+(* ******************************************************************************** *)
+scilla_version 0
+
+import IntUtils
+
+library Conversions
+
+type IntegerEncoding =
+  | LittleEndian
+  | BigEndian
+
+(* does not throw exceptions *)
+let osubstr : ByStr -> Uint32 -> Uint32 -> Option ByStr =
+  fun (bs : ByStr) =>
+  fun (pos : Uint32) =>
+  fun (len : Uint32) =>
+    let length = builtin strlen bs in
+    let shorter = uint32_le len length in
+    match shorter with
+    | True =>
+      let delta = builtin sub length len in
+      let safe_pos = uint32_le pos delta in
+      match safe_pos with
+      | True =>
+        let substr = builtin substr bs pos len in
+        Some {ByStr} substr
+      | False => None {ByStr}
+      end
+    | False => None {ByStr}
+    end
+
+(* Extract out a Scilla type from a ByStr starting at pos. Returns next position. *)
+(* Use the type specific helpers below rather than this function. *)
+let extract_scillaval : forall 'A. forall 'B. (ByStr -> Option 'B) -> ('B -> 'A) -> ByStr -> Uint32 -> Uint32 -> Option (Pair 'A Uint32) =
+  tfun 'A =>
+  tfun 'B =>
+  fun (to_bystrx : ByStr -> Option 'B) =>
+  fun (to_uint : 'B -> 'A) =>
+  fun (bs : ByStr) =>
+  fun (pos : Uint32) =>
+  fun (len : Uint32) =>         (* Byte size of the Uint value. *)
+    let subbs_opt =  osubstr bs pos len in
+    match subbs_opt with
+    | Some subbs =>
+      let subbs_x_opt = to_bystrx subbs in
+      match subbs_x_opt with
+      | Some subbs_x =>
```

Comment:
This pattern match an instance of `option_map : forall 'A. forall 'B. ('A -> 'B) -> Option 'A -> Option 'B`.

We currently don't have that function in the standard library, but we could add a file `OptionUtils.scillib` and fill it with the usual option type utilities from OCaml. Note that adding a new standard library is backward compatible, but adding new functions to an existing library is not, so if we go in that direction we should have a separate issue to create a "complete" set of option utliities, and then change Conversions.scillib to use this new library. (Note that adding `option_map` to Conversions.scillib is also fine, but removing it when we introduce OptionUtils is not.)

There are several other instances of `option_map` in this file, so I think it's probably worth doing something about it.

Reply:
I've tried this as a separtate PR against this PR branch. Please look at that and feel free to merge it (to this PR branch) if you see fit.